// azure-devops-ui/Components/IdentityCard/DefaultCard.js
import * as tslib_1 from "tslib";
import "../../CommonImports";
import "../../Core/core.css";
import "./IdentityCard.css";
import * as React from "react";
import * as Resources from '../../Resources.Persona';
import { Button } from '../../Button';
import { Icon } from '../../Icon';
import { Persona } from "../Persona/Persona";
import { CardContactLine } from "./CardContactLine";
var DefaultCard = /** @class */ (function (_super) {
    tslib_1.__extends(DefaultCard, _super);
    function DefaultCard() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // Setting the focus for the case of default card mounting when we come back via back header
    DefaultCard.prototype.componentDidMount = function () {
        if (this.props.setFocus) {
            this.props.setFocus();
        }
    };
    // Render
    DefaultCard.prototype.render = function () {
        var _this = this;
        var _a = this.props, identity = _a.identity, isPreviousHeader = _a.isPreviousHeader;
        var displayName = identity.displayName, department = identity.department, jobTitle = identity.jobTitle, mail = identity.mail, physicalDeliveryOfficeName = identity.physicalDeliveryOfficeName, telephoneNumber = identity.telephoneNumber;
        return (React.createElement("div", { className: isPreviousHeader ? "bolt-identity-default-card-with-header" : "bolt-identity-default-card-without-header" },
            React.createElement("div", { className: "flex-row" },
                React.createElement(Persona, { className: "bolt-identity-card-persona-main", size: 8 /* extraLarge */, identity: this.props.identity }),
                React.createElement("div", { className: "flex-column flex-grow bolt-identity-card-name" },
                    React.createElement("div", { className: "bolt-identity-card-persona-main-text-primary title-s" }, displayName),
                    jobTitle === "" ? undefined : React.createElement("div", { className: "bolt-identity-card-persona-main-text-secondary" }, jobTitle),
                    department === "" ? undefined : React.createElement("div", { className: "bolt-identity-card-persona-main-text-tertiary" }, department))),
            React.createElement("div", null,
                (mail || telephoneNumber || physicalDeliveryOfficeName) && (React.createElement("div", { className: "bolt-identity-default-card-info-wrapper " },
                    React.createElement("hr", { className: "bolt-identity-card-hr" }),
                    React.createElement("div", { className: "bolt-identity-default-card-header-wrapper" },
                        React.createElement(Button, { "aria-label": Resources.IdentityCardContact, className: "bolt-identity-default-card-header pointer bolt-profile-card-tab-element", onClick: function () {
                                _this.props.showContactCard();
                            }, role: "button", subtle: true, text: Resources.IdentityCardContact },
                            React.createElement(Icon, { iconName: "ChevronRight" }))),
                    React.createElement("div", { className: "bolt-identity-default-card-contact-info-container" },
                        mail && React.createElement(CardContactLine, { iconName: "Mail", content: mail, link: "mailto:" + mail }),
                        mail && React.createElement(CardContactLine, { iconName: "Chat", content: Resources.IdentityCardStartChat, link: "sip:" + mail }),
                        telephoneNumber && React.createElement(CardContactLine, { iconName: "Phone", content: telephoneNumber, link: "tel:" + telephoneNumber }),
                        physicalDeliveryOfficeName && React.createElement(CardContactLine, { iconName: "POI", content: physicalDeliveryOfficeName })))),
                this.renderDirectManagerElement())));
    };
    /**
     * Renders the direct manager for the persona.
     * @param managerIdentity The manager persona.
     */
    DefaultCard.prototype.renderDirectManagerElement = function () {
        var _this = this;
        if (!this.props.manager) {
            return React.createElement("div", null);
        }
        var adjacentManager = this.props.manager;
        var adjacentManagerData = {
            imageUrl: adjacentManager.image,
            primaryText: adjacentManager.displayName,
            secondaryText: adjacentManager.jobTitle,
            tertiaryText: adjacentManager.department
        };
        return (React.createElement("div", { className: "bolt-identity-default-card-direct-manager-wrapper" },
            React.createElement("hr", { className: "bolt-identity-card-hr" }),
            React.createElement("div", { className: "bolt-identity-default-card-header-wrapper" },
                React.createElement(Button, { "aria-label": Resources.IdentityCardOrganization, className: "bolt-identity-default-card-header pointer bolt-profile-card-tab-element", onClick: function () {
                        _this.props.showOrganizationCard && _this.props.showOrganizationCard();
                    }, role: "button", text: Resources.IdentityCardOrganization, subtle: true },
                    React.createElement(Icon, { iconName: "ChevronRight" }))),
            React.createElement("div", { className: "bolt-identity-default-card-direct-manager" },
                React.createElement("div", { className: "bolt-identity-default-card-header-reportsto" }, Resources.IdentityCardReportsTo),
                React.createElement("div", { className: "flex-row flex-grow" },
                    React.createElement(Button, { className: "bolt-bolt-identity-card-persona-list-manager-element bolt-profile-card-tab-element flex-row flex-grow", "aria-label": adjacentManager.displayName, key: adjacentManager.entityId + adjacentManager.signInAddress, onClick: function () {
                            _this.props.onClickEntity && _this.props.onClickEntity(adjacentManager);
                        }, subtle: true },
                        React.createElement(Persona, { size: 6 /* mediumPlus */, identity: adjacentManager }),
                        React.createElement("div", { className: "bolt-identity-card-text flex-column flex-grow " },
                            React.createElement("div", { className: "primary-text flex-row" }, adjacentManagerData.primaryText),
                            adjacentManagerData.secondaryText === "" ? (undefined) : (React.createElement("div", { className: "secondary-text flex-row" }, adjacentManagerData.secondaryText))))))));
    };
    return DefaultCard;
}(React.Component));
export { DefaultCard };
