// azure-devops-extension-api/Wiki/WikiClient.js
/*
 * ---------------------------------------------------------
 * Copyright(C) Microsoft Corporation. All rights reserved.
 * ---------------------------------------------------------
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
define(["require", "exports", "../Common/RestClientBase"], function (require, exports, RestClientBase_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var WikiRestClient = /** @class */ (function (_super) {
        __extends(WikiRestClient, _super);
        function WikiRestClient(options) {
            return _super.call(this, options) || this;
        }
        /**
         * Gets metadata or content of the wiki page for the provided path. Content negotiation is done based on the \`Accept\` header sent in the request.
         *
         * @param project - Project ID or project name
         * @param wikiIdentifier - Wiki Id or name.
         * @param path - Wiki page path.
         * @param recursionLevel - Recursion level for subpages retrieval. Defaults to \`None\` (Optional).
         * @param versionDescriptor - GitVersionDescriptor for the page. Defaults to the default branch (Optional).
         * @param includeContent - True to include the content of the page in the response for Json content type. Defaults to false (Optional)
         */
        WikiRestClient.prototype.getPageText = function (project, wikiIdentifier, path, recursionLevel, versionDescriptor, includeContent) {
            return __awaiter(this, void 0, void 0, function () {
                var queryValues;
                return __generator(this, function (_a) {
                    queryValues = {
                        path: path,
                        recursionLevel: recursionLevel,
                        versionDescriptor: versionDescriptor,
                        includeContent: includeContent
                    };
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.1",
                            httpResponseType: "text/plain",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}/pages/{*path}",
                            routeValues: {
                                project: project,
                                wikiIdentifier: wikiIdentifier
                            },
                            queryParams: queryValues
                        })];
                });
            });
        };
        /**
         * Gets metadata or content of the wiki page for the provided path. Content negotiation is done based on the \`Accept\` header sent in the request.
         *
         * @param project - Project ID or project name
         * @param wikiIdentifier - Wiki Id or name.
         * @param path - Wiki page path.
         * @param recursionLevel - Recursion level for subpages retrieval. Defaults to \`None\` (Optional).
         * @param versionDescriptor - GitVersionDescriptor for the page. Defaults to the default branch (Optional).
         * @param includeContent - True to include the content of the page in the response for Json content type. Defaults to false (Optional)
         */
        WikiRestClient.prototype.getPageZip = function (project, wikiIdentifier, path, recursionLevel, versionDescriptor, includeContent) {
            return __awaiter(this, void 0, void 0, function () {
                var queryValues;
                return __generator(this, function (_a) {
                    queryValues = {
                        path: path,
                        recursionLevel: recursionLevel,
                        versionDescriptor: versionDescriptor,
                        includeContent: includeContent
                    };
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.1",
                            httpResponseType: "application/zip",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}/pages/{*path}",
                            routeValues: {
                                project: project,
                                wikiIdentifier: wikiIdentifier
                            },
                            queryParams: queryValues
                        })];
                });
            });
        };
        /**
         * Gets metadata or content of the wiki page for the provided page id. Content negotiation is done based on the \`Accept\` header sent in the request.
         *
         * @param project - Project ID or project name
         * @param wikiIdentifier - Wiki Id or name.
         * @param id - Wiki page id.
         * @param recursionLevel - Recursion level for subpages retrieval. Defaults to \`None\` (Optional).
         * @param includeContent - True to include the content of the page in the response for Json content type. Defaults to false (Optional)
         */
        WikiRestClient.prototype.getPageByIdText = function (project, wikiIdentifier, id, recursionLevel, includeContent) {
            return __awaiter(this, void 0, void 0, function () {
                var queryValues;
                return __generator(this, function (_a) {
                    queryValues = {
                        recursionLevel: recursionLevel,
                        includeContent: includeContent
                    };
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.1",
                            httpResponseType: "text/plain",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}/pages/{id}",
                            routeValues: {
                                project: project,
                                wikiIdentifier: wikiIdentifier,
                                id: id
                            },
                            queryParams: queryValues
                        })];
                });
            });
        };
        /**
         * Gets metadata or content of the wiki page for the provided page id. Content negotiation is done based on the \`Accept\` header sent in the request.
         *
         * @param project - Project ID or project name
         * @param wikiIdentifier - Wiki Id or name.
         * @param id - Wiki page id.
         * @param recursionLevel - Recursion level for subpages retrieval. Defaults to \`None\` (Optional).
         * @param includeContent - True to include the content of the page in the response for Json content type. Defaults to false (Optional)
         */
        WikiRestClient.prototype.getPageByIdZip = function (project, wikiIdentifier, id, recursionLevel, includeContent) {
            return __awaiter(this, void 0, void 0, function () {
                var queryValues;
                return __generator(this, function (_a) {
                    queryValues = {
                        recursionLevel: recursionLevel,
                        includeContent: includeContent
                    };
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.1",
                            httpResponseType: "application/zip",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}/pages/{id}",
                            routeValues: {
                                project: project,
                                wikiIdentifier: wikiIdentifier,
                                id: id
                            },
                            queryParams: queryValues
                        })];
                });
            });
        };
        /**
         * Creates a new page view stats resource or updates an existing page view stats resource.
         *
         * @param project - Project ID or project name
         * @param wikiIdentifier - Wiki name or Id.
         * @param wikiVersion - Wiki version.
         * @param path - Wiki page path.
         * @param oldPath - Old page path. This is optional and required to rename path in existing page view stats.
         */
        WikiRestClient.prototype.createOrUpdatePageViewStats = function (project, wikiIdentifier, wikiVersion, path, oldPath) {
            return __awaiter(this, void 0, void 0, function () {
                var queryValues;
                return __generator(this, function (_a) {
                    queryValues = {
                        wikiVersion: wikiVersion,
                        path: path,
                        oldPath: oldPath
                    };
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.1",
                            method: "POST",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}/pageViewStats/{*path}",
                            routeValues: {
                                project: project,
                                wikiIdentifier: wikiIdentifier
                            },
                            queryParams: queryValues
                        })];
                });
            });
        };
        /**
         * Creates the wiki resource.
         *
         * @param wikiCreateParams - Parameters for the wiki creation.
         * @param project - Project ID or project name
         */
        WikiRestClient.prototype.createWiki = function (wikiCreateParams, project) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.2",
                            method: "POST",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}",
                            routeValues: {
                                project: project
                            },
                            body: wikiCreateParams
                        })];
                });
            });
        };
        /**
         * Deletes the wiki corresponding to the wiki name or Id provided.
         *
         * @param wikiIdentifier - Wiki name or Id.
         * @param project - Project ID or project name
         */
        WikiRestClient.prototype.deleteWiki = function (wikiIdentifier, project) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.2",
                            method: "DELETE",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}",
                            routeValues: {
                                project: project,
                                wikiIdentifier: wikiIdentifier
                            }
                        })];
                });
            });
        };
        /**
         * Gets all wikis in a project or collection.
         *
         * @param project - Project ID or project name
         */
        WikiRestClient.prototype.getAllWikis = function (project) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.2",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}",
                            routeValues: {
                                project: project
                            }
                        })];
                });
            });
        };
        /**
         * Gets the wiki corresponding to the wiki name or Id provided.
         *
         * @param wikiIdentifier - Wiki name or id.
         * @param project - Project ID or project name
         */
        WikiRestClient.prototype.getWiki = function (wikiIdentifier, project) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.2",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}",
                            routeValues: {
                                project: project,
                                wikiIdentifier: wikiIdentifier
                            }
                        })];
                });
            });
        };
        /**
         * Updates the wiki corresponding to the wiki Id or name provided using the update parameters.
         *
         * @param updateParameters - Update parameters.
         * @param wikiIdentifier - Wiki name or Id.
         * @param project - Project ID or project name
         */
        WikiRestClient.prototype.updateWiki = function (updateParameters, wikiIdentifier, project) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, this.beginRequest({
                            apiVersion: "5.1-preview.2",
                            method: "PATCH",
                            routeTemplate: "{project}/_apis/wiki/wikis/{wikiIdentifier}",
                            routeValues: {
                                project: project,
                                wikiIdentifier: wikiIdentifier
                            },
                            body: updateParameters
                        })];
                });
            });
        };
        WikiRestClient.RESOURCE_AREA_ID = "bf7d82a0-8aa5-4613-94ef-6172a5ea01f3";
        return WikiRestClient;
    }(RestClientBase_1.RestClientBase));
    exports.WikiRestClient = WikiRestClient;
});
