// azure-devops-extension-api/Dashboard/Dashboard.js
/*
 * ---------------------------------------------------------
 * Copyright(C) Microsoft Corporation. All rights reserved.
 * ---------------------------------------------------------
 */
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * identifies the scope of dashboard storage and permissions.
     */
    var DashboardScope;
    (function (DashboardScope) {
        DashboardScope[DashboardScope["Collection_User"] = 0] = "Collection_User";
        DashboardScope[DashboardScope["Project_Team"] = 1] = "Project_Team";
    })(DashboardScope = exports.DashboardScope || (exports.DashboardScope = {}));
    var GroupMemberPermission;
    (function (GroupMemberPermission) {
        GroupMemberPermission[GroupMemberPermission["None"] = 0] = "None";
        GroupMemberPermission[GroupMemberPermission["Edit"] = 1] = "Edit";
        GroupMemberPermission[GroupMemberPermission["Manage"] = 2] = "Manage";
        GroupMemberPermission[GroupMemberPermission["ManagePermissions"] = 3] = "ManagePermissions";
    })(GroupMemberPermission = exports.GroupMemberPermission || (exports.GroupMemberPermission = {}));
    var TeamDashboardPermission;
    (function (TeamDashboardPermission) {
        TeamDashboardPermission[TeamDashboardPermission["None"] = 0] = "None";
        TeamDashboardPermission[TeamDashboardPermission["Read"] = 1] = "Read";
        TeamDashboardPermission[TeamDashboardPermission["Create"] = 2] = "Create";
        TeamDashboardPermission[TeamDashboardPermission["Edit"] = 4] = "Edit";
        TeamDashboardPermission[TeamDashboardPermission["Delete"] = 8] = "Delete";
        TeamDashboardPermission[TeamDashboardPermission["ManagePermissions"] = 16] = "ManagePermissions";
    })(TeamDashboardPermission = exports.TeamDashboardPermission || (exports.TeamDashboardPermission = {}));
    /**
     * data contract required for the widget to function in a webaccess area or page.
     */
    var WidgetScope;
    (function (WidgetScope) {
        WidgetScope[WidgetScope["Collection_User"] = 0] = "Collection_User";
        WidgetScope[WidgetScope["Project_Team"] = 1] = "Project_Team";
    })(WidgetScope = exports.WidgetScope || (exports.WidgetScope = {}));
});
